File: 96. 不同的二叉搜索树/main.java
class Solution {
    public int numTrees(int n) {
        // 发现左右是子状态，因此可以用dp记录状态
        int[] dp = new int[n+1];
        dp[0] = 1;
        for (int i = 1; i <= n; i++) {
            dp[i] = 0;
            for (int j = 1; j <= i; j++) {
                dp[i] += dp[i-j] * dp[j-1];
            }
        }
        return dp[n];
    }
}

File: 96. 不同的二叉搜索树/main.go
func numTrees(n int) int {
    table := make([]int, n+2)
    table[0], table[1], table[2] = 1, 1, 2
    for i := 3; i <= n; i++ {
        for j := 1; j <= i; j++ {
			table[i] += table[j-1] * table[i-j]
		}
	}
	return table[n];
}
